worker/controlsocket/handlers.go

```diff
@@ -0,0 +1,237 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controlsocket
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/environs/bootstrap"
+	"github.com/juju/juju/state"
+)
+
+const (
+	// jujuMetricsUserPrefix defines the "namespace" in which this worker is
+	// allowed to create/remove users.
+	jujuMetricsUserPrefix = "juju-metrics-"
+
+	// userCreator is the listed "creator" of metrics users in state.
+	userCreator = "controlsocket"
+)
+
+func (w *Worker) registerHandlers(handle func(pattern string, handler http.Handler)) {
+	// We're using a fully REST-ful API structure here.
+	handle("/metrics-users", http.HandlerFunc(w.handleAddMetricsUser))
+	handle("/metrics-users/", http.HandlerFunc(w.handleRemoveMetricsUser))
+}
+
+func (w *Worker) handleAddMetricsUser(resp http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		resp.Header().Add("Allow", http.MethodPost)
+		w.writeResponse(resp, http.StatusMethodNotAllowed, errorResp("method must be %q, got %q", http.MethodPost, req.Method))
+		return
+	}
+
+	body := &struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+	}{}
+
+	defer req.Body.Close()
+	err := json.NewDecoder(req.Body).Decode(body)
+	if errors.Is(err, io.EOF) {
+		w.writeResponse(resp, http.StatusBadRequest, errorResp("missing request body"))
+		return
+	} else if err != nil {
+		w.writeResponse(resp, http.StatusBadRequest, errorResp("request body is not valid JSON: %v", err.Error()))
+		return
+	}
+
+	err = w.addMetricsUser(body.Username, body.Password)
+	if err != nil {
+		w.writeResponse(resp, asHTTPCode(err), errorResp(err.Error()))
+		return
+	}
+	w.writeResponse(resp, http.StatusOK, struct{}{})
+}
+
+func (w *Worker) addMetricsUser(username, password string) error {
+	err := validateMetricsUsername(username)
+	if err != nil {
+		return err
+	}
+
+	if password == "" {
+		return errors.NotValidf("empty password")
+	}
+
+	user, err := w.State.AddUser(username, username, password, userCreator)
+	if errors.Is(err, errors.AlreadyExists) {
+		if user.CreatedBy() != userCreator {
```

Comment:
If it already exists we don't update the password. This is supposed to be idempotent, so we need to confirm the password either matches or error.

Reply:
good point, and we should also check when removing it that it was created by the right user